staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,82 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"errors"
+	"math"
+)
+
+const maxHashBits = 64
+
+// ValidateParameters can validate parameters for shuffle sharding,
+// including queueSize and handSize
+// Algorithm: maxHashValue >= queueSize!/(queueSize-handSize)!
+func ValidateParameters(queueSize, handSize uint32) bool {
+	if handSize == 0 || queueSize == 0 || handSize > queueSize {
+		return false
+	}
+
+	maxHashValue := uint64(1<<maxHashBits - 1)
+	for i := uint32(0); i < handSize; i++ {
+		maxHashValue /= uint64(queueSize - i)
+		if maxHashValue == 0 {
+			return false
+		}
+	}
+
+	return true
+}
+
+// FastValidateParameters can validate parameters for shuffle sharding
+// in a fast but rough way, including queueSize and handSize
+// Algorithm: maxHashValue >= queueSize^handSize
+func FastValidateParameters(queueSize, handSize uint32) bool {
+	if handSize == 0 || queueSize == 0 || handSize > queueSize {
+		return false
+	}
+
+	return math.Log2(float64(queueSize))*float64(handSize) <= maxHashBits
+}
+
+// ShuffleSharding can shuffle a hash value to handSize-quantity and non-redundant
+// indices of queue
+// Eg. From queueSize=128, handSize=8, we can get an index array [12 14 73 18 119 51 117 26]
+// Algorithm: https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md#queue-assignment-proof-of-concept
+func ShuffleSharding(hashValue uint64, queueSize, handSize uint32) ([]uint32, error) {
+	if !ValidateParameters(queueSize, handSize) {
+		return nil, errors.New("bad parameters")
+	}
+
+	remainders, shuffledIndices := make([]uint32, handSize), make([]uint32, handSize)
+
+	for i := uint32(0); i < handSize; i++ {
+		remainders[i] = uint32(hashValue % uint64(queueSize-i))
+		hashValue /= uint64(queueSize - i)
```

Comment:
Note that the remainder operation (`%`) also does the same division.  This loop body could be re-written to do the division only once.

Reply:
OK, will update this.